File: GuideObjC/GuideObjC/ViewControllers/UserController.m
//
//  UserController.m
//  GuideObjC
//
//  Created by LuzanovRoman on 09.12.16.
//  Copyright © 2016 eigengraph. All rights reserved.
//

#import "UserController.h"

@interface UserController ()
@property (weak, nonatomic) IBOutlet UITextField *firstNameTextField;
@property (weak, nonatomic) IBOutlet UITextField *lastNameTextField;
@property (weak, nonatomic) IBOutlet UIButton *saveButton;
@property (retain, nonatomic) EGFUser *currentUser;
@property (retain, nonatomic) EGFUser *changedUser;
@end

@implementation UserController
@synthesize currentUser, changedUser;

- (void)viewDidLoad {
    [super viewDidLoad];
    [self.graph userObjectWithCompletion:^(NSObject * object, NSError * error) {
        if ([object isKindOfClass:[EGFUser class]]) {
            EGFUser * user = (EGFUser *)object;
            _firstNameTextField.text = user.name.given;
            _lastNameTextField.text = user.name.family;
            [self setCurrentUser:user];
        }
    }];
}

- (void)setCurrentUser:(EGFUser *)user {
    currentUser = user;
    changedUser = [user copyGraphObject];
    [self userDataDidUpdate];
}

- (IBAction)save:(id)sender {
    [self.view endEditing:true];
    
    [self.graph updateObjectWithId:currentUser.id object:changedUser completion:^(NSObject * object, NSError * error) {
        if ([object isKindOfClass:[EGFUser class]]) {
            [self setCurrentUser:(EGFUser *)object];
        }
    }];
}

- (IBAction)textDidChange:(id)sender {
    [self userDataDidUpdate];
}

- (void)userDataDidUpdate {
    if (!changedUser.name) {
        changedUser.name = [[EGFHumanName alloc] init];
    }
    changedUser.name.given = _firstNameTextField.text;
    changedUser.name.family = _lastNameTextField.text;
    _saveButton.enabled = ![currentUser isEqualWithGraphObject:changedUser];
}
@end

File: GuideSwift/GuideSwift/ViewControllers/UserController.swift
//
//  UserController.swift
//  GuideSwift
//
//  Created by LuzanovRoman on 09.12.16.
//  Copyright © 2016 eigengraph. All rights reserved.
//

import UIKit

class UserController: BaseController {
    
    @IBOutlet weak var firstNameTextField: UITextField!
    @IBOutlet weak var lastNameTextField: UITextField!
    @IBOutlet weak var saveButton: UIButton!
    fileprivate var currentUserObject: EGFUser?
    fileprivate var changedUserObject: EGFUser?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        Graph.userObject { (object, _) in
            guard let user = object as? EGFUser else { return }
            self.firstNameTextField.text = user.name?.given
            self.lastNameTextField.text = user.name?.family
            self.setCurrentUser(user)
        }
    }
    
    fileprivate func setCurrentUser(_ user: EGFUser) {
        currentUserObject = user
        changedUserObject = user.copyGraphObject()
        userDataDidUpdate()
    }
    
    @IBAction func save(_ sender: AnyObject) {
        guard let changedUser = changedUserObject, let id = changedUser.id else { return }
        view.endEditing(true)
        
        Graph.updateObject(withId: id, object: changedUser) { (object, error) in
            guard let user = object as? EGFUser else { return }
            self.setCurrentUser(user)
        }
    }
    
    @IBAction func textDidChange(_ sender: UITextField) {
        userDataDidUpdate()
    }
    
    fileprivate func userDataDidUpdate() {
        guard let currentUser = currentUserObject, let changedUser = changedUserObject else { return }
        
        if changedUser.name == nil {
            changedUser.name = EGFHumanName()
        }
        changedUser.name?.given = firstNameTextField.text
        changedUser.name?.family = lastNameTextField.text
        saveButton.isEnabled = !currentUser.isEqual(graphObject: changedUser)
    }
}

